How to define redirects in your repository (#592)
canonical/sphinx-docs-starter-pack: docs/how-to/optional-customisation/define-redirects.md

---
myst:
  html_meta:

>     description: How to define redirects in your repository

Reviewer: Required:

Per our meta description model, it should be something like this:

```suggestion
    description: How to redirect URLs in your documentation project. You can redirect a URL to a page inside your project, or to a page on another website.
```

The description shouldn't _replicate_ the title, otherwise the search result will be redundant:

> **Define redirects in your repository – Documentation starter pack documentation**
> How to define redirects in your repository

---

> % Original author: [s-makin](https://github.com/s-makin)

Reviewer: Nit:

I don't think this is necessary. It's all a group effort. :)

Author: Removed.

> # Define redirects in your repository

Reviewer: Recommended:

Should we consider a more succinct task-oriented title, like "Redirect URLs" or "Redirect pages"?

Reviewer 2: +1 to "Redirect pages" or "Redirect page URLs"

Author: Changed the title to Redirect pages

> If a file in your documentation set is moved, deleted or renamed, it can no longer be found at its original location and users will be shown a “404 Not Found” page. This can be frustrating for users.

Reviewer: Nit:

```suggestion
If a file in your documentation set is moved, deleted or renamed, it can no longer be found at its original location and users will be shown a 404 Not Found page. This can be frustrating for users.
```

I'm personally not a fan of quotation marks, and I think the capital letters are sufficient.

> If a file in your documentation set is moved, deleted or renamed, it can no longer be found at its original location and users will be shown a “404 Not Found” page. This can be frustrating for users.

Reviewer: Recommended:

I'd think we could this more straightforward:

```suggestion
Base Sphinx isn't capable of redirecting pages. If you move, delete, or rename a document's file, its URL becomes inaccessible.
```

Reviewer 2: Alternative suggestion:

```suggestion
If you move, delete, or rename a file, its original URL will lead to a 404 Not Found page.
```

"Base Sphinx" is a little confusing imo - it's a term we haven't defined or used often, so most people may not know what is meant by that.

I also think it's worth being explicit about the consequence being a 404 Not Found; paints a clearer picture of what the user will experience.

Reviewer: @a-velasco That's even better!

> To provide a better user experience, it is good practice to set up redirects to point to the new location, new file name, or to an alternative place where the information can be found. With documentation on Read the Docs, we can set up redirects [manually via the admin panel](https://docs.readthedocs.io/en/stable/guides/redirects.html). However, since this must be done separately, and *after* the file changes are made, this step can often be forgotten. 

Reviewer: Recommended:

> However, since this must be done separately, and *after* the file changes are made, this step can often be forgotten

It's possible to argue "_I_ would never forget" or "the workflow is fine for me and my small project."

I would instead say that RTD's workflow doesn't scale to complex and mature documentation.

> The **best practice** is to define redirects alongside the documentation in your repository, so that when a file is moved, renamed or deleted, you can create a redirect at the same time. In this way, redirects will always be in place when such changes go “live”, without needing to remember to do it (manually) afterwards. Redirects can either be **external**, meaning that they point to a URL outside your documentation, or **internal**, where you want to point to a new location within your documentation.

Reviewer: Recommended:

```suggestion
The **best practice** is to define and track redirects in your project source, so that when a file is moved, renamed or deleted, you can create a redirect at the same time. In this way, redirects will always be in place when such changes go live. Redirects can either be **external**, meaning that they point to a URL outside your documentation, or **internal**, where you want to point to a new location within your documentation.
```

> The **best practice** is to define redirects alongside the documentation in your repository, so that when a file is moved, renamed or deleted, you can create a redirect at the same time. In this way, redirects will always be in place when such changes go “live”, without needing to remember to do it (manually) afterwards. Redirects can either be **external**, meaning that they point to a URL outside your documentation, or **internal**, where you want to point to a new location within your documentation.

Reviewer: Another benefit of sourcing the redirects is that they are baked into the artifacts, making them portable. They work on the author's machine and when the docs are hosted on an airgapped network, and can be tested locally and with CI.

There are two ways to do this:

* For **external** and/or a **small number** of internal redirects, you can set up redirects in your project’s configuration file (`conf.py`) using *Reredirects*.  
* If you have many **internal** redirects, which is typical as documentation matures, you can avoid cluttering your configuration file by setting up *Rediraffe*, which conveniently lets you store your redirects in a separate file (`redirects.txt`).

> Note that if you need to specify external redirects *and* have a large number of internal redirects, it’s possible to use both methods at the same time.

Reviewer: This is completely solved by rerediraffe.

When migrating from Discourse to Read the Docs, the entire redirects table can be preserved in your `redirects.txt` file so that old/deep links from Discourse docs still redirect to the correct pages in the RTD documentation, and the web team can use a simple global redirect.

> ## Set up redirects in `conf.py`

Reviewer 3: The reredirects extension is still in the Starter Pack, but it's deprecated. We can still provide guidance on its usage, but the default should be to add all redirects, internal and external, with rediraffe.

Reviewer: Agreed. It or rerediraffe take one extra step to set up, but the benefits, and the benefits of everyone being on the same system, are significant.

@jahn-junior Is rerediraffe ready for wide adoption?

Author: I've removed everything about reredirects.

Reviewer: Sally's right about the scope – I was caught up in the fact that we have legacy support for rediraffe and reredirects. This page be solely focused on the default behaviour with re**re**diraffe.

> In the Starter Pack, the [reredirects Sphinx extension](https://documatt.com/sphinx-reredirects/usage/) is already installed. You can find the redirects section in your `conf.py` (or `custom_conf.py`) file, which looks something like this:

Reviewer: Required:

```suggestion
In the Starter Pack, the [reredirects Sphinx extension](https://documatt.com/sphinx-reredirects/usage/) is already installed. You can find the redirects section in your `conf.py` file, which looks something like this:
```

`custom_conf.py` is legacy and will be absent from any docs that updated to a recent version.

> ````

Reviewer: Out of curiosity, why four graves?

As a good practice, I also recommend using `code-block`, setting the language to Python, and using `:caption:` to name the file.

Author: Just how the automatic conversion by Gdoc worked. Updated.

#############
# Redirects #
#############

# To set up redirects: https://documatt.gitlab.io/sphinx-reredirects/usage.html
# For example: 'explanation/old-name.html': '../how-to/prettify.html',

# NOTE: If undefined, set to None, or empty,
#       the sphinx_reredirects extension will be disabled.

redirects = {}
````

> Any redirects you want to create are defined inside the “`redirects = {}`” variable. Inside this variable, add your redirects in the following patterns:

Reviewer: Required:

```suggestion
Add to the `redirects` variable, following these patterns:
```

````
redirects = {
	"old/path/": "https://redirect-to.com/new-url", # External
	"rename/oldname/": "rename/newname/", # Internal file rename
	"move/file/from/filename/": "../moved/to/filename/" # Move a file
}
````

> Here’s a [working example](https://github.com/canonical/ubuntu-server-documentation/blob/77d2718c06f5bcf504538cec09022131f9029a3d/custom_conf.py#L130):

Reviewer: Recommended:

It's redundant to provide a link and the block. I'm always looking for ways to reduce links. Can we cut it?

````
redirects = {
    "how-to/containers/lxc-containers": "https://linuxcontainers.org/lxc/documentation/",
    "reference/backups/basic-backup-shell-script": "https://discourse.ubuntu.com/t/basic-backup-shell-script/36419"
}
````

   

> See the [official reredirects documentation](https://documatt.com/sphinx-reredirects/usage/) for more information.

Reviewer: I'd link the project in `:relatedlinks:`

Author: I removed this one since I removed everything about reredirects but I added the Rediraffe docs link in `:relatedlinks:`

> ## Set up internal redirects (*in a separate file*)

Reviewer: Recommended:

I don't think this task name complements the previous task. The first is about doing it in `conf.py`, but this is primarily about internal redirects, which isn't a symmetrical choice for the reader. This is further reason IMO to remove any guidance about reredirects.

> To add the extension to your setup, add `sphinxext-rediraffe` in your `docs-requirements.txt` file. The file may be in the `.sphinx` subdirectory, and may just be called `requirements.txt`.

Reviewer: Required:

```suggestion
First, add `sphinxext-rediraffe` to your `requirements.txt` file.
```

Then, in your `conf.py` file, add the extension to the `extensions` list:

````
# Add to the list of extensions
extensions = [
	"sphinxext.rediraffe",
]
````

> Create a file called `redirects.txt` in the same directory as your `conf.py` (or `custom_conf.py`) file, on the branch that generates your documentation. This file will host all of your redirects.

Reviewer: Required:

```suggestion
Create a file called `redirects.txt` in the same directory as your `conf.py` file. This file will host all of your redirects.
```

I don't think discussing branching will help the user's work here. If there's something specific to branching, it should be its own decision or step.

> Next, in your `conf.py` (or `custom_conf.py`) file, add the `rediraffe_redirects` variable, which is the name of the redirects file you just created (include the file path if the file is not in the same directory as your conf.py):

Reviewer: Required:

```suggestion
Next, in your `conf.py` file, declare `rediraffe_redirects` variable and assign it the path of the redirects file you just created relative to `conf.py`:
```

````
# Add redirects, so they can be updated here to land alongside docs being moved
rediraffe_branch = "main"
rediraffe_redirects = "redirects.txt"
````

The format of the `redirects.txt` file is very specific. These instructions assume you are using `dirhtml` as your Sphinx builder (which avoids having .html at the end of your page URLs). If you are not using `dirhtml`, you should set this up by defining it in your `.readthedocs.yaml` file:

````
# Build documentation in the docs/ directory with Sphinx
sphinx:

>   builder: dirhtml

Reviewer: `builder: dirhtml` is default, so shouldn't we instead start this section with the advice that this extension won't work if the user has a different builder?

Author: Looking at the all [available builders](https://www.sphinx-doc.org/en/master/usage/builders/index.html), I wonder if there is a point in explaining this requirement in detail? It's not like it makes sense to have redirects for a single page HTML or a PDF file. I don't know all of the builders but at the first glance redirects are not relevant for other output formats by definition. IDK, maybe I'm overthinking it. :D 

Anyway, I moved it to be the first step in this section.

Reviewer: I don't think you're overthinking it. I don't think the extension actually has a negative effect on the default set up. Let's scrap this passage if it's redundant!

````

> ### Define redirects in the file

Reviewer: Recommended:

This is a separate and corollary task to setting up the extensions, so it should be an H2.

Inside your `redirects.txt` file, add your redirects in the following patterns:

````
# Client-side page redirects
# ---------------------------
# Comments start with a single hash (#)
# Each redirect appears on its own line in the format:
#
# redirect/from/ redirect/to/
#
# Paths are relative to the root of the `docs` directory
# The "from" path must be a file that *does not* exist
# The "to" path must be a file that *does* exist
# The separator between "from" and "to" paths is a single space (not a tab)

path/old-name/ path/new-name/			# To rename a file
path/file-name/ new-path/file-name/		# To move a file
path/old-name/ new-path/new-name/		# To move and rename a file
````

> Redirects are relative to the root of the `docs/` directory, so for simplicity, it’s best to put the redirects.txt file in the root directory. This is an [example of a working](https://github.com/canonical/ubuntu-server-documentation/blob/main/redirects.txt) `redirect.txt` file.

Reviewer: Recommended:

I'd move this into the extension setup task.

Author: I re-read it again and now I think we can get rid of it completely. We instruct specifically

>  Create a file called `redirects.txt` in the same directory as your `conf.py` file. This file will host all of your redirects.

So maybe we don't need to mention root at all? 

P.s. I also got rid of the example link as the file in the server docs was moved/deleted.

Reviewer: I agree, let's remove it!

### Layering redirects

In Rediraffe, multiple files can be redirected to the same target:

* A → D  
* B → D  
* C → D

> In your redirects.txt file, this looks like:

Reviewer: Nit:

```suggestion
In your `redirects.txt` file, this looks like:
```

````
installation tutorial/basic-installation/
basic-installation tutorial/basic-installation/
````

> This is common if files have been renamed more than once, creating layers of redirects. 

Reviewer: Great!
